base_layer/p2p/src/ping_pong.rs

```diff
@@ -0,0 +1,147 @@
+//  Copyright 2019 The Tari Project
+//
+//  Redistribution and use in source and binary forms, with or without modification, are permitted provided that the
+//  following conditions are met:
+//
+//  1. Redistributions of source code must retain the above copyright notice, this list of conditions and the following
+//  disclaimer.
+//
+//  2. Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the
+//  following disclaimer in the documentation and/or other materials provided with the distribution.
+//
+//  3. Neither the name of the copyright holder nor the names of its contributors may be used to endorse or promote
+//  products derived from this software without specific prior written permission.
+//
+//  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+//  INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+//  DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+//  SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+//  SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+//  WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE
+//  USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+use crate::{
+    services::{Service, ServiceContext, ServiceControlMessage},
+    tari_message::{NetMessage, TariMessageType},
+};
+use derive_error::Error;
+use log::*;
+use serde::{export::fmt::Debug, Deserialize, Serialize};
+use std::{sync::Arc, time::Duration};
+use tari_comms::{
+    domain_connector::ConnectorError,
+    message::{Message, MessageFlags, MessageHeader},
+    outbound_message_service::{outbound_message_service::OutboundMessageService, BroadcastStrategy},
+    DomainConnector,
+};
+use tari_utilities::message_format::MessageFormat;
+use threadpool::ThreadPool;
+
+const LOG_TARGET: &'static str = "base_layer::p2p::ping_pong";
+
+#[derive(Serialize, Deserialize)]
+pub enum PingPong {
+    Ping,
+    Pong,
+}
+
+#[derive(Debug, Error)]
+pub enum PingPongError {
+    ConnectorError(ConnectorError),
+}
+
+pub struct PingPongService {
+    interval: Duration,
+}
+
+fn format_err<E: Debug>(err: E) -> String {
+    format!("{:?}", err)
+}
+
+impl PingPongService {
+    pub fn new(interval: Duration) -> Self {
+        Self { interval }
+    }
+
+    fn run(&self, oms: &Arc<OutboundMessageService>, connector: &DomainConnector<'static>) -> Result<(), String> {
```

Comment:
Agree it looks a bit weird (also thinking using Arcs everywhere is probably an antipattern in some of the cases we're using them). As it stands, the OMS is shared between services, therefore the Arc - and the reference is simply to save creating another (unnecessary) Arc reference on every loop just so that we can OMS in other methods methods.

Reply:
`Arc::clone` is pretty cheap. 

I'm not saying drop the Arc, i'm saying choose between taking in an immutable borrow, or create a new atomic reference. YMMV but in `run` it looks like &OMS is fine (and remember, passing in an Arc<OMS> will also work because auto-deref)

Reply:
This is basically what I'm driving at:

https://gist.github.com/rust-play/e8dfcfd2946340df2a90cccacb4dadb3

Reply:
Ah thanks - that is cool!